sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,312 @@
+    return Cluster.builder()
+        .addContactPoints(hosts.toArray(new String[0]))
+        .withPort(port)
+        .build();
+  }
+
+  /**
+   * Gets the list of token ranges that a table occupies on a give Cassandra node.
+   *
+   * <p>NB: This method is compatible with Cassandra 2.1.5 and greater.
+   */
+  private static List<TokenRange> getTokenRanges(
+      Cluster cluster, String keyspace, String table) {
+    try (Session session = cluster.newSession()) {
+      ResultSet resultSet =
+          session.execute(
+              "SELECT range_start, range_end, partitions_count, mean_partition_size FROM "
+                  + "system.size_estimates WHERE keyspace_name = ? AND table_name = ?",
+              keyspace,
+              table);
+
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      for (Row row : resultSet) {
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getString("range_start"),
+                row.getString("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn(
+          "Failed to fetch size estimates for $keyspaceName.$tableName "
+              + "from system.size_estimates table. "
+              + "The number of created bundles may be inaccurate. "
+              + "Please make sure you use Cassandra 2.1.5 or newer.");
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      tokenRanges.add(new TokenRange(0L, 0L, "0", "0"));
+      return tokenRanges;
+    }
+  }
+
+  /**
+   * values that we get from system.size_estimates are for one node. We need to extrapolate to the
+   * whole cluster. This method estimates the percentage, the node represents in the cluster.
+   */
+  private static float getRingFraction(Cluster cluster, List<TokenRange> tokenRanges) {
+    Partitioner partitioner = new Partitioner(cluster);
+    long addressedTokens = 0;
+    for (TokenRange tokenRange : tokenRanges) {
+      addressedTokens =
+          addressedTokens + distance(cluster, tokenRange.rangeStart, tokenRange.rangeEnd);
+    }
+    // it is < 1 because it is a percentage
+    return Float.valueOf(addressedTokens / (partitioner.getTotalTokens()));
```

Comment:
This returned value will always be 0. Are there any tests for this method and for the methods that depend on it?